File: altgraph_web_app_java/src/main/java/org/cjoakim/cosmos/altgraph/data/common/graph/v2/struct/VertexValueStruct.java
package org.cjoakim.cosmos.altgraph.data.common.graph.v2.struct;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Data;
import org.cjoakim.cosmos.altgraph.data.common.graph.v2.JRank;
import org.cjoakim.cosmos.altgraph.data.common.graph.v2.JRankComparator;

import java.util.ArrayList;
import java.util.Collections;

@Data
public class VertexValueStruct {

    private long elapsedMs = -1;
    private String doctype = null;
    private String function = null;

    //private HashMap<String, Double> ranks;
    private ArrayList<JRank> ranks;

    public VertexValueStruct() {
        super();
        ranks = new ArrayList<JRank>();
        setDoctype("VertexValueStruct");
    }

    public void addRank(String vertex, Double value) {
        if (vertex != null) {
            ranks.add(new JRank(vertex, value));
        }
    }

    public JRank getRank(int idx) {
        return ranks.get(idx);
    }

    public void sort() {
        Collections.sort(ranks, new JRankComparator());
    }

    @JsonIgnore
    public String asJson(boolean pretty) throws Exception {

        try {
            ObjectMapper mapper = new ObjectMapper();
            if (pretty) {
                return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(this);
            } else {
                return mapper.writeValueAsString(this);
            }
        } catch (JsonProcessingException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: dotnet/src/altgraph_shared_app/Services/Graph/v2/Structs/VertexValueStruct.cs
using System.Text.Json.Serialization;

namespace altgraph_shared_app.Services.Graph.v2.Structs
{
  public class VertexValueStruct
  {
    public long ElapsedMs = -1;
    public string Doctype { get; private set; } = "VertexValueStruct";
    [JsonPropertyName("function")]
    public string Function { get; set; } = string.Empty;

    //public HashMap<string, Double> ranks;
    [JsonPropertyName("ranks")]
    public List<JRank> Ranks { get; private set; } = new List<JRank>();

    public VertexValueStruct()
    {
    }
    public VertexValueStruct(string function)
    {
      Function = function;
    }

    public void AddRank(string vertex, double value)
    {
      if (vertex != null)
      {
        Ranks.Add(new JRank(vertex, value));
      }
    }

    public JRank GetRank(int idx)
    {
      return Ranks[idx];
    }

    public void Sort()
    {
      Ranks.Sort(new JRankComparator());
    }
  }
}